tests/test_repository.py

```diff
@@ -237,6 +237,69 @@ def test_hg_log(fake_hg_repo):
     assert commits[2].node == revision4
 
 
+@responses.activate
+def test_download_component_mapping():
+    responses.add(
+        responses.HEAD,
+        "https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.source.source-bugzilla-info/artifacts/public/components.json",
+        status=200,
+        headers={"ETag": "100"},
+    )
+
+    responses.add(
+        responses.GET,
+        "https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.source.source-bugzilla-info/artifacts/public/components.json",
+        status=200,
+        json={},
+    )
+
+    repository.download_component_mapping()
+    assert len(repository.path_to_component) == 0
+
+    responses.reset()
+    responses.add(
+        responses.HEAD,
+        "https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.source.source-bugzilla-info/artifacts/public/components.json",
+        status=200,
+        headers={"ETag": "101"},
+    )
+
+    responses.add(
+        responses.GET,
+        "https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.source.source-bugzilla-info/artifacts/public/components.json",
+        status=200,
+        json={
+            "AUTHORS": ["mozilla.org", "Licensing"],
+            "Cargo.lock": ["Firefox Build System", "General"],
+        },
+    )
+
+    repository.download_component_mapping()
+    assert len(repository.path_to_component) == 2
+    assert repository.path_to_component["AUTHORS"] == "mozilla.org::Licensing"
+    assert repository.path_to_component["Cargo.lock"] == "Firefox Build System::General"
+
+    responses.reset()
+    responses.add(
+        responses.HEAD,
+        "https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.source.source-bugzilla-info/artifacts/public/components.json",
+        status=200,
+        headers={"ETag": "101"},
+    )
+
+    responses.add(
```

Comment:
This one could be removed (the code will check the etag and if it's the same it won't download the file).

Reply:
Ah, that's what my test case is, asserting that the response with the same etag won't be downloaded. If you think it's unnecessary, I will remove it when I have time later today.

Reply:
Exactly, so you just need to mock the HEAD request where the Etag is retrieved, this second request will never happen and so should not be mocked (this way we ensure it doesn't happen, which is what we want)

Reply:
Ah, I understand, I misunderstood and thought you meant the whole test case.